kong/db/dao/targets.lua

```diff
@@ -53,9 +53,7 @@ local function clean_history(self, upstream_pk)
     else
       -- haven't got this one, so this is the current state for this target
       seen[entry.target] = true
-      -- we can delete this entry if its weight == 0, but let's keep it if the
-      -- target list is almost empty, so the upstream still exists after cleaning up
-      if entry.weight == 0 and #targets > 2 then
+      if entry.weight == 0 then
```

Comment:
@locao instead of the old check, we could also do a check just a bit up, to set a minimum number of entries before ever starting a clean-up. Technically not needed, but would ease the "aggressiveness" earlier noted by @bungle Eg:

```lua
  if #target < 10 then
    -- too little entries, skip cleanup all together for efficiency reasons
    return
  end
```